crates/accelerate/src/gate_direction.rs

```diff
@@ -142,8 +157,505 @@ where
+
+    fix_gate_direction(py, dag, &coupling_map_check, None).cloned()
+}
+
+/// Try to swap two-qubit gate directions using pre-defined mapping to follow the right direction with respect to the given target.
+///
+/// Args:
+///     dag: the DAGCircuit to analyze
+///
+///     coupling_edges: set of edge pairs representing a directed coupling map, against which gate directionality is checked
+///
+/// Returns:
+///     the transformed DAGCircuit
+#[pyfunction]
+#[pyo3(name = "fix_gate_direction_target")]
+fn py_fix_with_target(py: Python, dag: &mut DAGCircuit, target: &Target) -> PyResult<DAGCircuit> {
+    let target_check = |inst: &PackedInstruction, op_args: &[Qubit]| -> bool {
+        let qargs = smallvec![
+            PhysicalQubit::new(op_args[0].0),
+            PhysicalQubit::new(op_args[1].0)
+        ];
+
+        // Take this path so Target can check for exact match of the parameterized gate's angle
+        if let OperationRef::Standard(std_gate) = inst.op.view() {
+            match std_gate {
+                StandardGate::RXXGate
+                | StandardGate::RYYGate
+                | StandardGate::RZZGate
+                | StandardGate::RZXGate => {
+                    return target
+                        .py_instruction_supported(
+                            py,
+                            None,
+                            Some(qargs),
+                            Some(
+                                get_std_gate_class(py, std_gate)
+                                    .expect("These gates should have Python classes")
+                                    .bind(py),
+                            ),
+                            Some(inst.params_view().to_vec()),
+                        )
+                        .unwrap_or(false)
+                }
+                _ => {}
+            }
+        }
+        target.instruction_supported(inst.op.name(), Some(&qargs))
+    };
+
+    fix_gate_direction(py, dag, &target_check, None).cloned()
+}
+
+// The main routine for fixing gate direction. Same parameters are check_gate_direction
+fn fix_gate_direction<'a, T>(
+    py: Python,
+    dag: &'a mut DAGCircuit,
+    gate_complies: &T,
+    qubit_mapping: Option<&[Qubit]>,
+) -> PyResult<&'a DAGCircuit>
+where
+    T: Fn(&PackedInstruction, &[Qubit]) -> bool,
+{
+    let mut nodes_to_replace: Vec<(NodeIndex, DAGCircuit)> = Vec::new();
+    let mut ops_to_replace: Vec<(NodeIndex, Vec<Bound<PyAny>>)> = Vec::new();
+
+    for node in dag.op_nodes(false) {
+        let NodeType::Operation(packed_inst) = &dag.dag()[node] else {
+            panic!("PackedInstruction is expected");
+        };
+
+        let op_args = dag.get_qargs(packed_inst.qubits);
+
+        if let OperationRef::Instruction(py_inst) = packed_inst.op.view() {
+            if py_inst.control_flow() {
+                let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py);
```

Comment:
For this I believe you can use the rust `circuit_to_dag` since it's not as lossy of a conversion. And since you're dealing with blocks you can keep the python `dag_to_circuit`. We'll improve on these structures in a follow-up :)

Reply:
Done in 4b363a19bd5920926d3f31748738f7b2070e4dd2